### PR TITLE
[CALCITE-1172] Add rule to flatten two Aggregate operators into one

### PR DESCRIPTION
**Pull request by hsyuan**
JIRA: https://issues.apache.org/jira/browse/CALCITE-1172

### CODE REVIEW

**Comment by hsyuan** [comment]
Can someone help review this pull request? thanks.

**Comment by vineetgarg02** [comment]
@hsyuan  Following query on an empty table should produce NULL
`select sum(y) from (
        select ename, count(mgr) y from 
        sales.emp group by ename, mgr) t`
But with this rule I believe this will be rewritten to
`select count(y) from t` which will produce 0.
Can you add a test case for this and verify?

Other similar query is:
`select sum(y) from (
        select ename, mgr, count(mgr) y from 
        sales.emp group by ename, mgr) t 
        group by cube(ename, mgr);`

This one should also output one row - NULL.

**Comment by hsyuan** [comment]
You are correct. The top aggregate must have group key, otherwise we can't merge them. Thanks for the case.

**Comment by hsyuan** [comment]
@vineetgarg02 I have addressed your comments. Please take a look.

File: core/src/main/java/org/apache/calcite/rel/rules/AggregateMergeRule.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    boolean hasEmptyGroup = topAgg.getGroupSets()
+        .stream().anyMatch(n -> n.isEmpty());
+
+    final List<AggregateCall> finalCalls = new ArrayList<>();
+    for (AggregateCall topCall : topAgg.getAggCallList()) {
+      if (!isAggregateSupported(topCall)
+          || topCall.getArgList().size() == 0) {
+        return;
+      }
+      int bottomIndex = topCall.getArgList().get(0) - bottomGroupSet.cardinality();
```
Comment (vineetgarg02) [inline core/src/main/java/org/apache/calcite/rel/rules/AggregateMergeRule.java]: A comment here explaining what this code is trying to achieve would be nice. e.g. `make sure top aggregate argument refers to one of the aggregate`

File: core/src/main/java/org/apache/calcite/sql/SqlSplittableAggFunction.java [hunk 2]
```diff
@@ -268,6 +303,19 @@ public AggregateCall topSplit(RexBuilder rexBuilder,
...
+          && (topKind == SqlKind.SUM
+              || topKind == SqlKind.SUM0)) {
+        return AggregateCall.create(bottom.getAggregation(), bottom.isDistinct(),
+            bottom.isApproximate(), bottom.getArgList(), bottom.filterArg,
+            bottom.getCollation(), bottom.getType(), top.getName());
+      } else {
+        return null;
+      }
+    }
+
```
Comment (vineetgarg02) [inline core/src/main/java/org/apache/calcite/sql/SqlSplittableAggFunction.java]: Another test case where both top and bottom aggregate has empty group by and are mergeable (e.g. SUM of SUM) would be nice.

**Comment by hsyuan** [comment]
@vineetgarg02 Thanks a lot. Will update the PR according.

**Comment by hsyuan** [comment]
@zhztheplayer Can you take a look at this PR? Need a committer to review it. Thanks :)

**Comment by zhztheplayer** [comment]
Oops I somehow missed your ping @hsyuan :(, reviewing now.

**Comment by zhztheplayer** [comment]
Overall I am +1, besides a minor issue (might not be an issue?):

If I test the code correctly, following SQL is still not optimized (bottom count has only an empty group set, and subquery is executed on empty table/set):

```
select sum(c) from (select count("empid") as c from "hr"."emps" where "deptno" < 0)
```

I guess this is because following fail condition is a little aggressive:
https://github.com/apache/calcite/blob/b5c6e25b131fcfcd5a640641401a8cb61820ad72/core/src/main/java/org/apache/calcite/rel/rules/AggregateMergeRule.java#L121-L123

Could we ensure the optimization happens on such a SQL too? (since it looks like a simplest one)

**Comment by hsyuan** [comment]
@zhztheplayer Good catch.  This is a special case that we can't just simply combining the 2 aggregates into a single aggregate, because the row type for top agg with SUM is integer nullable, but if we replace with COUNT, the row type becomes integer not nullable, we will see type mismatch assertion error. Unless we add a new Project operator on top to forcefully make it nullable, which seems like a hack.

**Comment by zhztheplayer** [comment]
Thank you for the explanation! I think It is indeed hacky to transform the COUNT to something nullable.  So I'm with you in general.

Maybe there is possibility to apply optimization to `SUM0(empty group)-COUNT(empty group)` relation, but the algebras seems to be rarely produced by SQL so it is likely not a very important case.

**Comment by hsyuan** [comment]
I am not sure whether it is worth doing for SUM0, because `SUM0` is not a standard sql aggregate function, and it only exists in Calcite.

**Comment by zhztheplayer** [comment]
> I am not sure whether it is worth doing for SUM0, because `SUM0` is not a standard sql aggregate function, and it only exists in Calcite.

Agreed. Thanks for the work, Haisheng!

**Comment by zhztheplayer** [comment]
@hsyuan - and looks like the PR is far behind current master so I suggest to rebase and trigger CI again.

**Comment by hsyuan** [comment]
@zhztheplayer Thanks! I have rebased on master and retriggered CI.
